### PR TITLE
Implement hrtime methods in Clock

### PR DESCRIPTION
**Pull request by KmeCnin**
I need to use `hrtime` to measure execution time and adding it to Clock allow us to enjoy good typing and mocked time.
I also add a bit of documentation, units disambiguation and fixed some other files for phpstan.

### CODE REVIEW

File: src/Clock/Clock.php [hunk 1]
```diff
@@ -55,9 +75,13 @@ public static function time(): int
     }
 
     /**
-     * @return int|false
```
Comment (KmeCnin) [inline src/Clock/Clock.php:58]: Since PHP 8, sleep is not returning `false` anymore and throws instead.

File: src/Pcre/PcreUtils.php [hunk 2]
```diff
@@ -15,6 +15,7 @@ class PcreUtils
      * @see https://www.php.net/manual/en/function.preg-match.php
      *
      * @param array<string> $matches
+     * @param int<0|256|512|768> $flags
```
Comment (KmeCnin) [inline src/Pcre/PcreUtils.php:18]: File updated to match phpstan complaints

File: src/Pcre/PcreUtils.php [hunk 3]
```diff
@@ -15,6 +15,7 @@ class PcreUtils
      * @see https://www.php.net/manual/en/function.preg-match.php
      *
      * @param array<string> $matches
+     * @param 0|256|512|768 $flags
```
Comment (KmeCnin) [inline src/Pcre/PcreUtils.php]: File updated to match phpstan complaints

**Review by cajax — APPROVED** [review]

**Review by treuliaux — APPROVED** [review]

File: src/Clock/Clock.php (hunk 1 above)
Comment (arnaud-lb) [reply src/Clock/Clock.php:58]: We could actually improve the interface by forbidding negative delays (sleep returns false / throws when passing a negative delay).

File: src/Clock/Clock.php [hunk 4]
```diff
@@ -14,19 +16,27 @@
  */
 class Clock
 {
-    /** @var ?float */
-    private static $now;
+    /** @var ?float number of seconds or null if not mocked */
```
Comment (arnaud-lb) [inline src/Clock/Clock.php]: Nit pick: We could specify that this is the number of seconds _since the Unix Epoch (January 1 1970 00:00:00 GMT)_

File: src/Pcre/PcreUtils.php (hunk 3 above)
Comment (arnaud-lb) [reply src/Pcre/PcreUtils.php]: We may be able to use the `int-mask-of<...>` type here

File: src/Clock/Clock.php [hunk 5]
```diff
@@ -76,6 +119,83 @@ public static function usleep(int $us): void
...
+            return hrtime(false);
+        }
+
+        $seconds = (int) self::$now;
+
+        return [
+            // Whole seconds
+            $seconds,
+            // Remaining nanoseconds
+            (int) round((self::$now - $seconds) * 1e+9),
```
Comment (arnaud-lb) [inline src/Clock/Clock.php:139]: We could avoid the cast with `* 1_000_000_000`

Edit: actually not, since self::$now is a float

**Review by arnaud-lb — COMMENTED** [review]
Nice!

File: src/Pcre/PcreUtils.php (hunk 3 above)
Comment (KmeCnin) [reply src/Pcre/PcreUtils.php]: stylé !

**Pull request merged**